### PR TITLE
add a storage client for boltdb-shipper which would do all the object key management for storage operations

### PR DESCRIPTION
**Pull request by sandeepsukhani**
**What this PR does / why we need it**:
The code for managing object keys for boltdb files created by `boltdb-shipper` is spread at various places. This PR adds a storage client which would take care of managing the keys and provides a simplified interface for various object storage operations for managing index files in the long term storage.

**Checklist**
- [x] Tests updated

### CODE REVIEW

File: pkg/storage/stores/shipper/storage/client.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	tableNames := make([]string, 0, len(tables))
+	for _, table := range tables {
+		tableNames = append(tableNames, path.Base(string(table)))
+	}
+
+	return tableNames, nil
+}
+
+func (s *indexStorageClient) ListFiles(ctx context.Context, tableName string) ([]IndexFile, error) {
+	// The forward slash here needs to stay because we are trying to list contents of a directory without which we will get the name of the same directory back with hosted object stores.
```
Comment (owen-d) [inline pkg/storage/stores/shipper/storage/client.go:54]: tiny nit: it helps to keep these comment lines a bit shorter so we don't have to scroll :)

**Review by owen-d — APPROVED** [review]
Mostly skimmed, but a nice refactor PR. Thanks <3

File: pkg/storage/stores/shipper/storage/client.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+type Client interface {
+	ListTables(ctx context.Context) ([]string, error)
+	ListFiles(ctx context.Context, tableName string) ([]IndexFile, error)
+	GetFile(ctx context.Context, tableName, fileName string) (io.ReadCloser, error)
+	PutFile(ctx context.Context, tableName, fileName string, file io.ReadSeeker) error
+	DeleteFile(ctx context.Context, tableName, fileName string) error
+	IsObjectNotFoundErr(err error) bool
+	Stop()
+}
```
Comment (jeschkies) [inline pkg/storage/stores/shipper/storage/client.go:23]: I'm not sure this is important but should this interface be documented?

**Review by jeschkies — APPROVED** [review]
Thanks. This looks good. I just have a question on documentation.

File: pkg/storage/stores/shipper/storage/client.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+const delimiter = "/"
+
+type Client interface {
+	ListTables(ctx context.Context) ([]string, error)
+	ListFiles(ctx context.Context, tableName string) ([]IndexFile, error)
+	GetFile(ctx context.Context, tableName, fileName string) (io.ReadCloser, error)
+	PutFile(ctx context.Context, tableName, fileName string, file io.ReadSeeker) error
+	DeleteFile(ctx context.Context, tableName, fileName string) error
+	IsObjectNotFoundErr(err error) bool
```
Comment (slim-bean) [inline pkg/storage/stores/shipper/storage/client.go:21]: I am wondering if this should be `IsObjectNotFound` the signature makes it obvious it takes an error type.

When I first saw this I was confused what exactly this method was doing, I think removing the `Err` from the name makes it more clear to me

**Comment by sandeepsukhani** [comment]
Thanks, folks for the review! I have taken care of the feedback. Merging this.

**Pull request merged**